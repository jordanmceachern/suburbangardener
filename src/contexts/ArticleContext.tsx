"use client";

import React, {
  createContext,
  useContext,
  useReducer,
  useCallback,
} from "react";
import type { ArticleDisplayData } from "../components/articles/ArticleDisplayCard";

interface ArticleState {
  articles: ArticleDisplayData[];
  featuredArticle: ArticleDisplayData | null;
  total: number;
  hasMore: boolean;
  loading: boolean;
  error: string | null;
  searchTerm: string;
  selectedCategory: string;
  currentOffset: number;
}

type ArticleAction =
  | { type: "SET_LOADING"; payload: boolean }
  | { type: "SET_ERROR"; payload: string | null }
  | {
      type: "SET_ARTICLES";
      payload: {
        articles: ArticleDisplayData[];
        total: number;
        hasMore: boolean;
      };
    }
  | {
      type: "APPEND_ARTICLES";
      payload: { articles: ArticleDisplayData[]; hasMore: boolean };
    }
  | { type: "SET_FEATURED_ARTICLE"; payload: ArticleDisplayData | null }
  | { type: "SET_SEARCH_TERM"; payload: string }
  | { type: "SET_SELECTED_CATEGORY"; payload: string }
  | { type: "SET_OFFSET"; payload: number }
  | { type: "RESET_PAGINATION" };

const initialState: ArticleState = {
  articles: [],
  featuredArticle: null,
  total: 0,
  hasMore: false,
  loading: false,
  error: null,
  searchTerm: "",
  selectedCategory: "All Categories",
  currentOffset: 0,
};

function articleReducer(
  state: ArticleState,
  action: ArticleAction
): ArticleState {
  switch (action.type) {
    case "SET_LOADING":
      return { ...state, loading: action.payload };
    case "SET_ERROR":
      return { ...state, error: action.payload, loading: false };
    case "SET_ARTICLES":
      return {
        ...state,
        articles: action.payload.articles,
        total: action.payload.total,
        hasMore: action.payload.hasMore,
        loading: false,
        error: null,
        currentOffset: action.payload.articles.length,
      };
    case "APPEND_ARTICLES":
      return {
        ...state,
        articles: [...state.articles, ...action.payload.articles],
        hasMore: action.payload.hasMore,
        loading: false,
        currentOffset: state.currentOffset + action.payload.articles.length,
      };
    case "SET_FEATURED_ARTICLE":
      return { ...state, featuredArticle: action.payload };
    case "SET_SEARCH_TERM":
      return { ...state, searchTerm: action.payload };
    case "SET_SELECTED_CATEGORY":
      return { ...state, selectedCategory: action.payload };
    case "SET_OFFSET":
      return { ...state, currentOffset: action.payload };
    case "RESET_PAGINATION":
      return { ...state, currentOffset: 0, hasMore: false };
    default:
      return state;
  }
}

interface ArticleContextType {
  state: ArticleState;
  loadArticles: () => Promise<void>;
  loadMoreArticles: () => Promise<void>;
  searchArticles: (searchTerm: string) => Promise<void>;
  setSelectedCategory: (category: string) => void;
  getFilteredArticles: () => ArticleDisplayData[];
}

const ArticleContext = createContext<ArticleContextType | null>(null);

export function useArticles() {
  const context = useContext(ArticleContext);
  if (!context) {
    throw new Error("useArticles must be used within an ArticleProvider");
  }
  return context;
}

interface ArticleProviderProps {
  children: React.ReactNode;
}

export function ArticleProvider({ children }: ArticleProviderProps) {
  const [state, dispatch] = useReducer(articleReducer, initialState);

  const loadArticles = useCallback(
    async (reset = false) => {
      dispatch({ type: "SET_LOADING", payload: true });
      dispatch({ type: "SET_ERROR", payload: null });

      if (reset) {
        dispatch({ type: "RESET_PAGINATION" });
      }

      try {
        const offset = reset ? 0 : state.currentOffset;
        const params = new URLSearchParams({
          limit: "20",
          offset: offset.toString(),
          ...(state.searchTerm && { search: state.searchTerm }),
        });

        const response = await fetch(`/api/articles?${params}`);
        if (!response.ok) {
          throw new Error("Failed to fetch articles");
        }

        const data = await response.json();

        if (reset || offset === 0) {
          dispatch({
            type: "SET_ARTICLES",
            payload: {
              articles: data.articles,
              total: data.total,
              hasMore: data.hasMore,
            },
          });

          // Set featured article (first article)
          if (data.articles.length > 0) {
            dispatch({
              type: "SET_FEATURED_ARTICLE",
              payload: data.articles[0],
            });
          }
        } else {
          dispatch({
            type: "APPEND_ARTICLES",
            payload: {
              articles: data.articles,
              hasMore: data.hasMore,
            },
          });
        }
      } catch (error) {
        dispatch({
          type: "SET_ERROR",
          payload: error instanceof Error ? error.message : "An error occurred",
        });
      }
    },
    [state.currentOffset, state.searchTerm]
  );

  const loadMoreArticles = useCallback(async () => {
    if (!state.hasMore || state.loading) return;
    await loadArticles(false);
  }, [state.hasMore, state.loading, loadArticles]);

  const searchArticles = useCallback(
    async (searchTerm: string) => {
      dispatch({ type: "SET_SEARCH_TERM", payload: searchTerm });
      await loadArticles(true);
    },
    [loadArticles]
  );

  const setSelectedCategory = useCallback((category: string) => {
    dispatch({ type: "SET_SELECTED_CATEGORY", payload: category });
  }, []);

  const getFilteredArticles = useCallback(() => {
    let filtered = state.articles;

    // Exclude featured article from the list to avoid duplication
    if (state.featuredArticle) {
      filtered = filtered.filter(
        article => article.id !== state.featuredArticle!.id
      );
    }

    // Apply category filter (client-side filtering)
    if (state.selectedCategory !== "All Categories") {
      filtered = filtered.filter(
        article => article.category === state.selectedCategory
      );
    }

    return filtered;
  }, [state.articles, state.selectedCategory, state.featuredArticle]);

  const contextValue: ArticleContextType = {
    state,
    loadArticles,
    loadMoreArticles,
    searchArticles,
    setSelectedCategory,
    getFilteredArticles,
  };

  return (
    <ArticleContext.Provider value={contextValue}>
      {children}
    </ArticleContext.Provider>
  );
}
