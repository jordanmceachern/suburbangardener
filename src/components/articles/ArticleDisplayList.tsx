"use client";

import React from "react";
import Image from "next/image";
import ArticleDisplayCard, {
  type ArticleDisplayData,
} from "./ArticleDisplayCard";
import { useArticles } from "../../contexts/ArticleContext";

interface ArticleDisplayListProps {
  className?: string;
}

export default function ArticleDisplayList({
  className = "",
}: ArticleDisplayListProps) {
  const { state, loadMoreArticles, getFilteredArticles } = useArticles();

  // Get articles filtered by category (client-side filtering)
  const displayedArticles = getFilteredArticles();

  if (state.loading && state.articles.length === 0) {
    return (
      <div className={`${className}`}>
        <div className="grid grid-cols-1 md:grid-cols-2 xl:grid-cols-3 gap-6">
          {Array.from({ length: 6 }).map((_, index) => (
            <div key={index} className="animate-pulse">
              <div className="bg-neutral-200 dark:bg-neutral-700 rounded-lg h-48 sm:h-64 mb-4" />
              <div className="bg-neutral-200 dark:bg-neutral-700 rounded h-6 mb-2" />
              <div className="bg-neutral-200 dark:bg-neutral-700 rounded h-4 mb-2" />
              <div className="bg-neutral-200 dark:bg-neutral-700 rounded h-4 w-2/3" />
            </div>
          ))}
        </div>
      </div>
    );
  }

  return (
    <div className={`${className}`}>
      {/* No Articles Message */}
      {!state.loading && displayedArticles.length === 0 && (
        <div className="text-center py-12">
          {/* eslint-disable-next-line jsx-a11y/alt-text -- alt text is provided via label prop */}
          <Image
            alt="a ruined squished tomato on the ground"
            className="flex justify-center mb-8 rounded-lg mx-auto"
            height={400}
            width={400}
            src="https://res.cloudinary.com/dtweazqf2/image/upload/q_auto,f_auto/v1755004794/SuburbanGardener/toflato_l0dkqy.jpg"
          />
          <p className="text-xl text-neutral-900 dark:text-neutral-100">
            Sorry! We encountered an issue while trying to fetch our articles
            for you, please come back to check again later.
          </p>
        </div>
      )}

      {/* Articles Grid */}
      {displayedArticles.length > 0 && (
        <div className="grid grid-cols-1 md:grid-cols-2 xl:grid-cols-3 gap-6">
          {displayedArticles.map(article => (
            <ArticleDisplayCard key={article.id} article={article} />
          ))}
        </div>
      )}

      {/* Load More Articles */}
      {state.hasMore && (
        <div className="mt-8 text-center">
          <button
            onClick={loadMoreArticles}
            disabled={state.loading}
            className="bg-primary-600 hover:bg-primary-700 disabled:bg-primary-400 text-white px-8 py-3 rounded-lg font-medium transition-colors inline-flex items-center gap-2"
          >
            {state.loading ? (
              <>
                <svg
                  className="animate-spin h-5 w-5"
                  fill="none"
                  viewBox="0 0 24 24"
                >
                  <circle
                    className="opacity-25"
                    cx="12"
                    cy="12"
                    r="10"
                    stroke="currentColor"
                    strokeWidth="4"
                  />
                  <path
                    className="opacity-75"
                    fill="currentColor"
                    d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
                  />
                </svg>
                Loading...
              </>
            ) : (
              `Load More Articles`
            )}
          </button>
        </div>
      )}

      {/* Loading More Skeleton */}
      {state.loading && state.articles.length > 0 && (
        <div className="grid grid-cols-1 md:grid-cols-2 xl:grid-cols-3 gap-6 mt-6">
          {Array.from({ length: 3 }).map((_, index) => (
            <div key={`loading-${index}`} className="animate-pulse">
              <div className="bg-neutral-200 dark:bg-neutral-700 rounded-lg h-48 sm:h-64 mb-4" />
              <div className="bg-neutral-200 dark:bg-neutral-700 rounded h-6 mb-2" />
              <div className="bg-neutral-200 dark:bg-neutral-700 rounded h-4 mb-2" />
              <div className="bg-neutral-200 dark:bg-neutral-700 rounded h-4 w-2/3" />
            </div>
          ))}
        </div>
      )}

      {state.error && (
        <div className="mt-8 text-center">
          <p className="text-red-600 dark:text-red-400">{state.error}</p>
        </div>
      )}
    </div>
  );
}
