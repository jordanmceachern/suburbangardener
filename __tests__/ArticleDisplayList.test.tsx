import { render } from "@testing-library/react";
import ArticleDisplayList from "@/components/articles/ArticleDisplayList";
import { ArticleProvider } from "@/contexts/ArticleContext";

describe("ArticleDisplayList Component", () => {
  const mockProps = {
    className: "test-list-class",
  };

  it("should match DOM snapshot", () => {
    const { container } = render(
      <ArticleProvider>
        <ArticleDisplayList {...mockProps} />
      </ArticleProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should maintain CSS classes for list layout", () => {
    const { container } = render(
      <ArticleProvider>
        <ArticleDisplayList {...mockProps} />
      </ArticleProvider>
    );

    // Check for list-related classes
    const hasListClasses =
      container.innerHTML.includes("grid") ||
      container.innerHTML.includes("flex") ||
      container.innerHTML.includes("space-") ||
      container.innerHTML.includes("gap-");

    expect(hasListClasses).toBe(true);
  });
});
