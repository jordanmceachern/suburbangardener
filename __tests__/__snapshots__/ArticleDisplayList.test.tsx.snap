// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ArticleDisplayList Component should match DOM snapshot 1`] = `
<div
  class="test-list-class"
>
  <div
    class="text-center py-12"
  >
    <img
      alt="a ruined squished tomato on the ground"
      class="flex justify-center mb-8 rounded-lg mx-auto"
      data-nimg="1"
      decoding="async"
      height="400"
      loading="lazy"
      src="/_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdtweazqf2%2Fimage%2Fupload%2Fq_auto%2Cf_auto%2Fv1755004794%2FSuburbanGardener%2Ftoflato_l0dkqy.jpg&w=828&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdtweazqf2%2Fimage%2Fupload%2Fq_auto%2Cf_auto%2Fv1755004794%2FSuburbanGardener%2Ftoflato_l0dkqy.jpg&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdtweazqf2%2Fimage%2Fupload%2Fq_auto%2Cf_auto%2Fv1755004794%2FSuburbanGardener%2Ftoflato_l0dkqy.jpg&w=828&q=75 2x"
      style="color: transparent;"
      width="400"
    />
    <p
      class="text-xl text-neutral-900 dark:text-neutral-100"
    >
      Sorry! We encountered an issue while trying to fetch our articles for you, please come back to check again later.
    </p>
  </div>
</div>
`;
